apps/api-extractor/src/generators/PackageTypingsGenerator.ts

```diff
@@ -0,0 +1,402 @@
+        case ts.SyntaxKind.EnumKeyword:
+        case ts.SyntaxKind.NamespaceKeyword:
+        case ts.SyntaxKind.ModuleKeyword:
+        case ts.SyntaxKind.TypeKeyword:
+          span.modification.prefix = 'declare ' + span.modification.prefix;
+          if (entry.exported) {
+            span.modification.prefix = 'export ' + span.modification.prefix;
+          }
+          break;
+
+        case ts.SyntaxKind.VariableDeclaration:
+          if (!parentSpan) {
+            // The VariableDeclaration node is part of a VariableDeclarationList, however
+            // the Entry.followedSymbol points to the VariableDeclaration part because
+            // multiple definitions might share the same VariableDeclarationList.
+            //
+            // Since we are emitting a separate declaration for each one, we need to look upwards
+            // in the ts.Node tree and write a copy of the enclosing VariableDeclarationList
+            // content (e.g. "var" from "var x=1, y=2").
+            const list: ts.VariableDeclarationList | undefined = PackageTypingsGenerator._matchParent(span.node,
+              [ts.SyntaxKind.VariableDeclaration, ts.SyntaxKind.VariableDeclarationList]);
+            if (!list) {
+              throw new Error('Unsupported variable declaration');
+            }
+            const listPrefix: string = list.getSourceFile().text
+              .substring(list.getStart(), list.declarations[0].getStart());
+            span.modification.prefix = 'declare ' + listPrefix + span.modification.prefix;
+            span.modification.suffix = ';';
+          }
+          break;
+
+        case ts.SyntaxKind.Identifier:
+          if (parentSpan) {
+            switch (parentSpan.kind) {
+              case ts.SyntaxKind.ExpressionWithTypeArguments:
+              case ts.SyntaxKind.TypeReference:
+
+              case ts.SyntaxKind.ClassDeclaration:
+              case ts.SyntaxKind.InterfaceDeclaration:
+              case ts.SyntaxKind.EnumDeclaration:
+              case ts.SyntaxKind.TypeAliasDeclaration:
+              case ts.SyntaxKind.ModuleDeclaration:  // (namespaces are a type of module declaration)
+                {
+                  const symbol: ts.Symbol | undefined = this._typeChecker.getSymbolAtLocation(span.node);
+                  if (!symbol) {
+                    throw new Error('Symbol not found');
+                  }
+
+                  const referencedEntry: Entry | undefined = this._fetchEntryForSymbol(symbol);
+                  if (referencedEntry) {
+                    if (!referencedEntry.uniqueName) {
+                      // This should never happen
+                      throw new Error('referencedEntry.uniqueName is undefined');
+                    }
+
+                    span.modification.prefix = referencedEntry.uniqueName;
+                    // span.modification.prefix += '/**/';
+                  }
+                }
+                break;
+            }
+          }
```

Comment:
Describe why this is necessary.

Reply:
Let's consider _modifySpan() a prototype for now.  I need to get it wired up and see how it works for various real world inputs first, then I'll come back and rework this code.  Some of the logic will be completely eliminated when we start processing *.d.ts inputs instead of *.ts.

---
In reply to: [152427693](https://github.com/Microsoft/web-build-tools/pull/448#discussion_r152427693) [](ancestors = 152427693)